File: monitors/rppg.py
"""
rPPG — ประเมินอัตราการเต้นหัวใจจากกล้อง (แบบไม่ต้องแตะตัว)

ใช้อัลกอริทึม POS (Plane-Orthogonal-to-Skin, Wang et al. 2017) ซึ่งเป็น
gold-standard ของ rPPG: รวมสัญญาณทั้ง 3 สี (R,G,B) แล้ว project ลงระนาบ
ที่ตั้งฉากกับการเปลี่ยนแปลงของโทนผิว ทำให้ทนต่อการขยับหน้า/แสงเปลี่ยน
ได้ดีกว่าวิธีดูช่องเขียวอย่างเดียวมาก (วิธีเขียวล้วนมักล็อกความถี่ต่ำผิด ๆ
ทำให้ค่าชีพจรออกมาต่ำเกินจริง)

ขั้นตอน: เก็บค่าเฉลี่ย R,G,B ของ ROI หน้าผากตามเวลา -> POS ได้คลื่นชีพจร
-> bandpass ในย่านชีพจรมนุษย์ -> FFT หาความถี่เด่น -> แปลงเป็น bpm
นอกจากนี้ยังเปิดคลื่นชีพจรที่กรองแล้ว (self.filtered, self.fs) ให้
StressMonitor นำไปคำนวณ HRV ต่อ
"""
from collections import deque

import numpy as np
from scipy.signal import butter, filtfilt

# ย่านชีพจรมนุษย์: 0.75-4.0 Hz = 45-240 ครั้ง/นาที
# (ขยับขอบล่างจาก 0.7 ขึ้นมาเล็กน้อย กันไปจับ component ช้าจากการหายใจ/ขยับ)
_LOW_HZ = 0.75
_HIGH_HZ = 4.0


class RPPGEstimator:
    def __init__(self, window_seconds: float = 10.0, min_seconds: float = 6.0):
        self.window_seconds = window_seconds
        self.min_seconds = min_seconds
        self._times: deque[float] = deque()
        self._r: deque[float] = deque()
        self._g: deque[float] = deque()
        self._b: deque[float] = deque()

        self.bpm: float | None = None
        self.confidence: float = 0.0
        # คลื่นชีพจรที่กรองแล้ว + sampling rate ล่าสุด (ให้ StressMonitor ใช้)
        self.filtered: np.ndarray | None = None
        self.fs: float | None = None

    def update(self, roi_bgr, timestamp: float) -> None:
        """ป้อนภาพ ROI หน้าผาก (BGR) เข้ามาทีละเฟรม"""
        if roi_bgr is None or roi_bgr.size == 0:
            return
        # OpenCV เป็น BGR: ช่อง 0=B, 1=G, 2=R
        self._b.append(float(np.mean(roi_bgr[:, :, 0])))
        self._g.append(float(np.mean(roi_bgr[:, :, 1])))
        self._r.append(float(np.mean(roi_bgr[:, :, 2])))
        self._times.append(timestamp)

        while self._times and (timestamp - self._times[0]) > self.window_seconds:
            self._times.popleft()
            self._r.popleft()
            self._g.popleft()
            self._b.popleft()

    def _pos(self, rgb: np.ndarray, fs: float) -> np.ndarray | None:
        """
        POS algorithm: rgb เป็น (N,3) ของ R,G,B คืนคลื่นชีพจรความยาว N
        ใช้หน้าต่างเลื่อน 1.6 วินาที + overlap-add ตามต้นฉบับ
        """
        n_total = len(rgb)
        win = int(1.6 * fs)
        if win < 2 or n_total < win:
            return None

        signal = np.zeros(n_total)
        chans = rgb.T  # (3, N): แถว 0=R, 1=G, 2=B
        for start in range(0, n_total - win + 1):
            end = start + win
            block = chans[:, start:end]
            mu = block.mean(axis=1, keepdims=True)
            normalized = block / (mu + 1e-9)        # temporal normalization
            s1 = normalized[1] - normalized[2]                       # G - B
            s2 = normalized[1] + normalized[2] - 2 * normalized[0]   # G + B - 2R
            alpha = np.std(s1) / (np.std(s2) + 1e-9)
            pulse = s1 + alpha * s2
            signal[start:end] += pulse - pulse.mean()  # overlap-add
        return signal

    def estimate(self) -> float | None:
        """คืนค่า BPM ถ้าข้อมูลพอ ไม่งั้นคืน None"""
        n = len(self._times)
        if n < 30:
            return None
        duration = self._times[-1] - self._times[0]
        if duration < self.min_seconds:
            return None
        fs = n / duration               # เฟรมเรตจริง (ไม่คงที่)
        if fs < 5:
            return None

        rgb = np.stack(
            [np.asarray(self._r), np.asarray(self._g), np.asarray(self._b)], axis=1)
        pulse = self._pos(rgb, fs)
        if pulse is None:
            return None

        nyq = fs / 2.0
        try:
            b, a = butter(3, [_LOW_HZ / nyq, _HIGH_HZ / nyq], btype="band")
            filtered = filtfilt(b, a, pulse)
        except Exception:
            return None

        self.filtered = filtered
        self.fs = fs

        windowed = filtered * np.hanning(len(filtered))
        spectrum = np.abs(np.fft.rfft(windowed))
        freqs = np.fft.rfftfreq(len(windowed), d=1.0 / fs)
        band = (freqs >= _LOW_HZ) & (freqs <= _HIGH_HZ)
        if not np.any(band):
            return None

        band_power = spectrum[band]
        band_freqs = freqs[band]
        peak_idx = int(np.argmax(band_power))
        bpm = float(band_freqs[peak_idx] * 60.0)

        total = float(np.sum(band_power))
        self.confidence = float(band_power[peak_idx] / total) if total > 0 else 0.0
        self.bpm = bpm if self.bpm is None else (0.8 * self.bpm + 0.2 * bpm)
        return self.bpm

File: web/src/monitors/rppg.ts
/**
 * rPPG — วัดอัตราการเต้นหัวใจจากกล้อง (พอร์ตจาก monitors/rppg.py)
 *
 * ใช้อัลกอริทึม POS (Plane-Orthogonal-to-Skin, Wang 2017): รวมสัญญาณ R/G/B
 * ของ ROI หน้าผาก แล้ว project ลงระนาบตั้งฉากกับโทนผิว → คลื่นชีพจร
 * จากนั้น bandpass (ทำผ่าน FFT/IFFT แทน scipy.filtfilt) → หา peak ใน FFT
 *
 * เปิด this.filtered / this.fs (คลื่นชีพจรที่กรองแล้ว) ให้ StressMonitor (HRV)
 * นำไปใช้ในเฟสถัดไป
 */
import FFT from "fft.js";

const LOW_HZ = 0.75; // 45 bpm
const HIGH_HZ = 4.0; // 240 bpm

function nextPow2(n: number): number {
  let p = 1;
  while (p < n) p <<= 1;
  return p;
}

function std(a: number[]): number {
  const n = a.length;
  if (!n) return 0;
  const mean = a.reduce((s, v) => s + v, 0) / n;
  const varr = a.reduce((s, v) => s + (v - mean) * (v - mean), 0) / n;
  return Math.sqrt(varr);
}

function hann(n: number): Float64Array {
  const w = new Float64Array(n);
  for (let i = 0; i < n; i++) w[i] = 0.5 - 0.5 * Math.cos((2 * Math.PI * i) / (n - 1));
  return w;
}

/** bandpass แบบ zero-phase: FFT → ตัด bin นอกย่าน → IFFT (แทน filtfilt) */
function bandpassTime(signal: number[], fs: number, low: number, high: number): Float64Array {
  const N = nextPow2(signal.length);
  const fft = new FFT(N);
  const input = new Array(N).fill(0);
  for (let i = 0; i < signal.length; i++) input[i] = signal[i];
  const spec = fft.createComplexArray();
  fft.realTransform(spec, input);
  fft.completeSpectrum(spec);
  for (let k = 0; k < N; k++) {
    const f = (k <= N / 2 ? k : k - N) * (fs / N); // ความถี่แบบมีเครื่องหมาย
    const af = Math.abs(f);
    if (af < low || af > high) {
      spec[2 * k] = 0;
      spec[2 * k + 1] = 0;
    }
  }
  const out = fft.createComplexArray();
  fft.inverseTransform(out, spec);
  const res = new Float64Array(signal.length);
  for (let i = 0; i < signal.length; i++) res[i] = out[2 * i]; // เอาเฉพาะส่วนจริง
  return res;
}

export class RPPGEstimator {
  private times: number[] = [];
  private r: number[] = [];
  private g: number[] = [];
  private b: number[] = [];

  bpm: number | null = null;
  confidence = 0;
  filtered: Float64Array | null = null;
  fs: number | null = null;

  constructor(private windowSeconds = 10, private minSeconds = 6) {}

  /** ป้อนค่าเฉลี่ย R/G/B ของ ROI หน้าผาก + เวลา (วินาที) */
  update(rMean: number, gMean: number, bMean: number, tSec: number): void {
    this.r.push(rMean);
    this.g.push(gMean);
    this.b.push(bMean);
    this.times.push(tSec);
    while (this.times.length && tSec - this.times[0] > this.windowSeconds) {
      this.times.shift();
      this.r.shift();
      this.g.shift();
      this.b.shift();
    }
  }

  /** POS: คืนคลื่นชีพจรความยาวเท่าจำนวนเฟรม (หน้าต่างเลื่อน 1.6s + overlap-add) */
  private pos(fs: number): number[] | null {
    const n = this.times.length;
    const win = Math.floor(1.6 * fs);
    if (win < 2 || n < win) return null;

    const signal = new Array(n).fill(0);
    for (let start = 0; start <= n - win; start++) {
      let mr = 0, mg = 0, mb = 0;
      for (let i = start; i < start + win; i++) {
        mr += this.r[i];
        mg += this.g[i];
        mb += this.b[i];
      }
      mr /= win; mg /= win; mb /= win;

      const s1 = new Array(win);
      const s2 = new Array(win);
      for (let i = 0; i < win; i++) {
        const cr = this.r[start + i] / (mr + 1e-9);
        const cg = this.g[start + i] / (mg + 1e-9);
        const cb = this.b[start + i] / (mb + 1e-9);
        s1[i] = cg - cb;
        s2[i] = cg + cb - 2 * cr;
      }
      const alpha = std(s1) / (std(s2) + 1e-9);
      let hmean = 0;
      const h = new Array(win);
      for (let i = 0; i < win; i++) {
        h[i] = s1[i] + alpha * s2[i];
        hmean += h[i];
      }
      hmean /= win;
      for (let i = 0; i < win; i++) signal[start + i] += h[i] - hmean; // overlap-add
    }
    return signal;
  }

  estimate(): number | null {
    const n = this.times.length;
    if (n < 30) return null;
    const duration = this.times[n - 1] - this.times[0];
    if (duration < this.minSeconds) return null;
    const fs = n / duration;
    if (fs < 5) return null;

    const pulse = this.pos(fs);
    if (!pulse) return null;

    // bandpass = เทียบเท่า filtfilt; เก็บไว้ให้ HRV ใช้ต่อ
    const filtered = bandpassTime(pulse, fs, LOW_HZ, HIGH_HZ);
    this.filtered = filtered;
    this.fs = fs;

    // FFT บนสัญญาณที่กรอง + Hann window → หา peak ในย่านชีพจร
    const w = hann(filtered.length);
    const windowed = Array.from(filtered, (v, i) => v * w[i]);
    const N = nextPow2(windowed.length);
    const fft = new FFT(N);
    const input = new Array(N).fill(0);
    for (let i = 0; i < windowed.length; i++) input[i] = windowed[i];
    const spec = fft.createComplexArray();
    fft.realTransform(spec, input);
    fft.completeSpectrum(spec);

    let peakIdx = -1;
    let peakVal = -1;
    let total = 0;
    for (let k = 0; k <= N / 2; k++) {
      const freq = (k * fs) / N;
      if (freq < LOW_HZ || freq > HIGH_HZ) continue;
      const mag = Math.hypot(spec[2 * k], spec[2 * k + 1]);
      total += mag;
      if (mag > peakVal) {
        peakVal = mag;
        peakIdx = k;
      }
    }
    if (peakIdx < 0) return null;

    const bpm = ((peakIdx * fs) / N) * 60;
    this.confidence = total > 0 ? peakVal / total : 0;
    this.bpm = this.bpm == null ? bpm : 0.8 * this.bpm + 0.2 * bpm;
    return this.bpm;
  }
}
